File: 25-ReverseNodesInKGroup/25-ReverseNodesInKGroup.cpp
// Last updated: 6/25/2026, 9:17:58 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(k==1) return head;
        int length = 0; ListNode* trav = head;
        while(trav != NULL) {trav = trav->next; length++;}
        ListNode* post; trav=head; ListNode* prev=NULL;
        for(int i=0; i<length/k; i++){
            int t=k-1; ListNode* first = trav; ListNode* postnext;
            post = trav->next;
            while(t){
                postnext = post->next;
                post->next = trav;
                trav = post;
                post = postnext;
                t--;
            }
            if(prev==NULL) head = trav;
            else{prev->next = trav;}
            first->next = post;
            prev = first;
            trav = prev->next;
        }
        return head;
    }
};

File: 25-ReverseNodesInKGroup/25-ReverseNodesInKGroup.py
# Last updated: 6/25/2026, 9:17:55 AM
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: Optional[ListNode]
        :type k: int
        :rtype: Optional[ListNode]
        """
        curr1=ListNode(0,head)
        prev1=curr1
        curr=head
        while curr:
            k_node=self.group(curr,k)
            if not k_node:
                break
            k_nxt=k_node.next
            new_head=self.reverse(curr,k_node.next,k_nxt)
            prev1.next=new_head
            prev1=curr
            curr=k_nxt

        return curr1.next
    
    def reverse(self,head,prev,k_next):
        if  head==k_next:
            return prev
        nxt=head.next
        head.next=prev
        return self.reverse(nxt,head,k_next)
    
    def group(self,head,k):
        while head and k>1:
            head=head.next
            k-=1
        return head